stl/inc/stop_token

```diff
@@ -0,0 +1,401 @@
+
+            _Head->_Fn(_Head); // might destroy *_Head
+        }
+    }
+};
+
+class stop_source;
+
+class stop_token {
+    friend stop_source;
+    friend _Stop_callback_base;
+
+public:
+    stop_token() noexcept : _State{} {}
+    stop_token(const stop_token& _Other) noexcept : _State{_Other._State} {
+        const auto _Local = _State;
+        if (_Local != nullptr) {
+            _Local->_Stop_tokens.fetch_add(1, memory_order_relaxed);
+        }
+    }
+
+    stop_token(stop_token&& _Other) noexcept : _State{_STD exchange(_Other._State, nullptr)} {}
+    stop_token& operator=(const stop_token& _Other) noexcept {
+        stop_token{_Other}.swap(*this);
+        return *this;
+    }
+
+    stop_token& operator=(stop_token&& _Other) noexcept {
+        stop_token{_STD move(_Other)}.swap(*this);
+        return *this;
+    }
+
+    ~stop_token() {
+        const auto _Local = _State;
+        if (_Local != nullptr) {
+            if (_Local->_Stop_tokens.fetch_sub(1, memory_order_acq_rel) == 1) {
+                delete _Local;
+            }
+        }
+    }
+
+    void swap(stop_token& _Other) noexcept {
+        _STD swap(_State, _Other._State);
+    }
+
+    [[nodiscard]] bool stop_requested() const noexcept {
+        const auto _Local = _State;
+        return _Local != nullptr && _Local->_Stop_requested();
+    }
+
+    [[nodiscard]] bool stop_possible() const noexcept {
+        const auto _Local = _State;
+        return _Local != nullptr && _Local->_Stop_possible();
+    }
+
+    [[nodiscard]] friend bool operator==(const stop_token& _Lhs, const stop_token& _Rhs) noexcept {
+        return _Lhs._State == _Rhs._State;
+    }
+
+    friend void swap(stop_token& _Lhs, stop_token& _Rhs) noexcept {
+        _STD swap(_Lhs._State, _Rhs._State);
+    }
+
+private:
+    explicit stop_token(_Stop_state* const _State_) : _State{_State_} {}
+
+    _Stop_state* _State;
+};
+
+class stop_source {
+public:
+    stop_source() : _State{new _Stop_state} {}
+    explicit stop_source(nostopstate_t) noexcept : _State{} {}
+    stop_source(const stop_source& _Other) noexcept : _State{_Other._State} {
+        const auto _Local = _State;
+        if (_Local != nullptr) {
+            _Local->_Stop_sources.fetch_add(2, memory_order_relaxed);
+        }
+    }
+
+    stop_source(stop_source&& _Other) noexcept : _State{_STD exchange(_Other._State, nullptr)} {}
+    stop_source& operator=(const stop_source& _Other) noexcept {
+        stop_source{_Other}.swap(*this);
+        return *this;
+    }
+
+    stop_source& operator=(stop_source&& _Other) noexcept {
+        stop_source{_STD move(_Other)}.swap(*this);
+        return *this;
+    }
+
+    ~stop_source() {
+        const auto _Local = _State;
+        if (_Local != nullptr) {
+            if ((_Local->_Stop_sources.fetch_sub(2, memory_order_relaxed) >> 1) == 1) {
+                if (_Local->_Stop_tokens.fetch_sub(1, memory_order_acq_rel) == 1) {
+                    delete _Local;

```

Comment:
```suggestion
            if (_Local->_Stop_sources.fetch_sub(2, memory_order_release) == (1 << 1)) {
                if (_Local->_Stop_tokens.fetch_sub(1, memory_order_release) == 1) {
                    atomic_thread_fence(memory_order_acquire);
                    delete _Local;
```

Reply:
Hmmmm I'm 'scared' enough about explicit fences and ensuring they always do the right thing, and this case is very rare, that I prefer to just make them both acq_rel.